### PR TITLE
🌱 test/e2e: Add v1.0=>main clusterctl upgrade test

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #8038

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,6 +14,7 @@ hack/tools/bin
 test/e2e/data/infrastructure-docker/main/cluster-template*.yaml
 test/e2e/data/infrastructure-docker/v0.3/cluster-template*.yaml
 test/e2e/data/infrastructure-docker/v0.4/cluster-template*.yaml
+test/e2e/data/infrastructure-docker/v1.0/cluster-template*.yaml
```
Comment (jackfrancis) [inline .gitignore:17]: would it be best to make this more general going forward so we didn't have to manually maintain this with every release? e.g. `test/e2e/data/infrastructure-docker/**/cluster-template*.yaml`

Ref:

- https://git-scm.com/docs/gitignore

**Comment by sbueringer** [comment]
test ?

File: .gitignore (hunk 1 above)
Comment (sbueringer) [reply .gitignore:17]: Yup, thx!

File: test/e2e/clusterctl_upgrade_test.go [hunk 2]
```diff
@@ -58,6 +58,30 @@ var _ = Describe("When testing clusterctl upgrades (v0.4=>current)", func() {
...
+			ArtifactFolder:        artifactFolder,
+			SkipCleanup:           skipCleanup,
+			InitWithBinary:        "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.0.5/clusterctl-{OS}-{ARCH}",
+			// We have to pin the providers because with `InitWithProvidersContract` the test would
+			// use the latest version for the contract (which is v1.3.0 for v1beta1).
+			InitWithCoreProvider:            "cluster-api:v1.0.5",
+			InitWithBootstrapProviders:      []string{"kubeadm:v1.0.5"},
+			InitWithControlPlaneProviders:   []string{"kubeadm:v1.0.5"},
+			InitWithInfrastructureProviders: []string{"docker:v1.0.5"},
+			// We have to set this to an empty array as clusterctl v1.2 doesn't support
```
Comment (jackfrancis) [inline test/e2e/clusterctl_upgrade_test.go:76]: nit: copy pasta should read:

```
// We have to set this to an empty array as clusterctl v1.0 doesn't support
```

(I assume we will be using the same version of clusterctl as the capi cluster version we're building)

**Comment by sbueringer** [comment]
WIP => Have to trial-and-error e2e tests a bit

/test pull-cluster-api-e2e-full-main

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by sbueringer** [comment]
Given lgtm above + it's a straightforward change 

/approve

**Pull request merged**